### PR TITLE
[Mystikos] Support the CapturePFGPExceptions preference

### PR DESCRIPTION
**Pull request by mingweishih**
This PR does the following to the mystikos.v4 branch
- Cherry pick #3984 to the mystikos branch.
- Address the problem with exception simulation (https://github.com/deislabs/mystikos/issues/399)
- Update the pf_gp_exception test
- Fix a bug in `OE_ENCLAVE_FLAG_DEBUG_AUTO`

Signed-off-by: Ming-Wei Shih <mishih@microsoft.com>

### CODE REVIEW

File: enclave/core/sgx/exception.c [hunk 1]
```diff
@@ -446,6 +457,17 @@ void oe_virtual_exception_dispatcher(
     }
     else
     {
+        /* The following codes can only be captured with SGX2 and the
+         * MISCSELECT[0] bit is set to 1. */
+        if (is_exit_info_valid &&
```
Comment (jxyang) [inline enclave/core/sgx/exception.c:462]: Where do we check `MISCSELECT[0] == 1`?

File: tests/pf_gp_exceptions/host/host.c [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+static bool _is_misc_region_supported()
+{
+    uint32_t eax, ebx, ecx, edx;
+    eax = ebx = ecx = edx = 0;
+
+    // Obtain feature information using CPUID
+    oe_get_cpuid(CPUID_SGX_LEAF, 0x0, &eax, &ebx, &ecx, &edx);
+
+    // Check if EXINFO is supported by the processor
+    if (!(ebx & CPUID_SGX_MISC_EXINFO_MASK))
```
Comment (jxyang) [inline tests/pf_gp_exceptions/host/host.c:22]: Just use `return (ebx & CPUID_SGX_MISC_EXINFO_MASK);`

File: tests/pf_gp_exceptions/enc/enc.c [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+uint64_t test_pfgp_handler(oe_exception_record_t* exception_record)
+{
+    if (exception_record->code == OE_EXCEPTION_PAGE_FAULT)
+    {
+        const int MOV_INSTRUCTION_BYTES = 3;
+        faulting_address = exception_record->faulting_address;
+        error_code = exception_record->error_code;
+        exception_record->context->rip += MOV_INSTRUCTION_BYTES;
+        oe_host_printf(
```
Comment (jxyang) [inline tests/pf_gp_exceptions/enc/enc.c:20]: It's unsafe to print any message inside an exception handler.

If we really need it, save the variables and print them after the control is transferred from the exception handler to the regular code.

File: tests/pf_gp_exceptions/enc/enc.c [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    /* Trigger #PF */
+    faulting_address = 1;
+    asm volatile("xor %rax, %rax\n\t"
+                 "mov (%rax), %rax");
+    OE_TEST(faulting_address == 0);
+
+    /* Trigger #GP */
+    faulting_address = 1;
+    asm volatile("lgdt 0x00");
+    OE_TEST(faulting_address == 0);
```
Comment (jxyang) [inline tests/pf_gp_exceptions/enc/enc.c:64]: Verify the error code?

File: tests/pf_gp_exceptions/enc/enc.c [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    if (result != OE_OK)
+    {
+        return -1;
+    }
+
+    /* Trigger #PF */
+    faulting_address = 1;
+    asm volatile("xor %rax, %rax\n\t"
+                 "mov (%rax), %rax");
+    OE_TEST(faulting_address == 0);
```
Comment (jxyang) [inline tests/pf_gp_exceptions/enc/enc.c:59]: Is it possible to test #PF triggered by page permission?

File: enclave/core/sgx/exception.c (hunk 1 above)
Comment (jxyang) [reply enclave/core/sgx/exception.c:462]: I was curious if this branch should only be activated when `MISCSELECT[0] bit is set to 1`, why that is not in the if-condition of the branch?

File: enclave/core/sgx/exception.c [hunk 6]
```diff
@@ -385,10 +394,12 @@ void oe_virtual_exception_dispatcher(
...
     }
-    sgx_ssa_gpr_t* ssa_gpr =
-        (sgx_ssa_gpr_t*)(((uint8_t*)ssa_info.base_address) + ssa_info.frame_byte_size - OE_SGX_GPR_BYTE_SIZE);
+    uint64_t gprsgx_offset = (uint64_t)ssa_info.base_address +
+                             ssa_info.frame_byte_size - OE_SGX_GPR_BYTE_SIZE;
+    sgx_ssa_gpr_t* ssa_gpr = (sgx_ssa_gpr_t*)gprsgx_offset;
+    bool is_exit_info_valid = ssa_gpr->exit_info.as_fields.valid;
 
-    if (!ssa_gpr->exit_info.as_fields.valid)
+    if (!is_exit_info_valid)
```
Comment (jxyang) [inline enclave/core/sgx/exception.c:402]: The following code is unsafe. We should strength the condition to: if (oe_exception_record && debug_mode)...
```c
// Consider manually provided exception information for SGX1
        if (oe_exception_record)
        {
            td->exception_address = ssa_gpr->rip;

            /* TODO PRP: We need to read the exception information from
             * oe_exception_record instead */
            td->exception_code = OE_EXCEPTION_PAGE_FAULT;
            td->exception_flags |= OE_EXCEPTION_FLAGS_HARDWARE;
        }
```
This is issue https://github.com/deislabs/mystikos/issues/399

Comment (mingweishih) [reply enclave/core/sgx/exception.c:402]: Thanks for the catch. I actually have another PR pending for this fix.
https://github.com/mingweishih/openenclave/commit/453da6c18a967e961c1e3d728eeb27a0e38e2953

I can add the fix to this PR as well.

File: enclave/core/sgx/exception.c (hunk 1 above)
Comment (mingweishih) [reply enclave/core/sgx/exception.c:462]: ah, I misunderstood your question. When `MISCSELECT[0] bit is set to 1`, it is SGX hardware that decides to capture #PF/#GP; i.e., if we configure the enclave with `MISCSELECT[0] bit set to 0`, the #PF/#GP will never be delivered to the enclave. Therefore, we don't need additional software check.

File: tests/pf_gp_exceptions/enc/enc.c (hunk 5 above)
Comment (mingweishih) [reply tests/pf_gp_exceptions/enc/enc.c:59]: Will update on this.

File: tests/pf_gp_exceptions/enc/enc.c (hunk 4 above)
Comment (mingweishih) [reply tests/pf_gp_exceptions/enc/enc.c:64]: Will update this, thanks.

File: enclave/core/sgx/exception.c (hunk 6 above)
Comment (mingweishih) [reply enclave/core/sgx/exception.c:403]: Updated, please check.

File: tests/pf_gp_exceptions/enc/enc.c (hunk 4 above)
Comment (mingweishih) [reply tests/pf_gp_exceptions/enc/enc.c]: Updated.

File: tests/pf_gp_exceptions/enc/enc.c (hunk 5 above)
Comment (mingweishih) [reply tests/pf_gp_exceptions/enc/enc.c]: For simplicity, I keep the use of address 0, which is unmapped or inaccessible on both Linux and Windows by default.

File: tests/pf_gp_exceptions/enc/enc.c (hunk 3 above)
Comment (mingweishih) [reply tests/pf_gp_exceptions/enc/enc.c]: removed the print

File: tests/pf_gp_exceptions/host/host.c (hunk 2 above)
Comment (mingweishih) [reply tests/pf_gp_exceptions/host/host.c]: done

**Review by jxyang — APPROVED** [review]

File: enclave/core/sgx/exception.c [hunk 7]
```diff
@@ -385,20 +395,28 @@ void oe_virtual_exception_dispatcher(
...
             td->exception_address = ssa_gpr->rip;
 
             /* TODO PRP: We need to read the exception information from
              * oe_exception_record instead */
             td->exception_code = OE_EXCEPTION_PAGE_FAULT;
             td->exception_flags |= OE_EXCEPTION_FLAGS_HARDWARE;
+
+            /* We cannot know the faulting address that triggers #PF on SGX1
+             * unless the host passes in the information. */
+            td->faulting_address = 0;
```
Comment (bodzhang) [inline enclave/core/sgx/exception.c:418]: Why can't the host pass in the faulting_address info?

File: enclave/core/sgx/exception.c [hunk 8]
```diff
@@ -385,20 +395,28 @@ void oe_virtual_exception_dispatcher(
...
+    if (!is_exit_info_valid)
     {
         // Consider manually provided exception information for SGX1
-        if (oe_exception_record)
+        // Only allow the simulated #PF in debug mode
+        if (is_enclave_debug_allowed() && oe_exception_record)
         {
             td->exception_address = ssa_gpr->rip;
 
             /* TODO PRP: We need to read the exception information from
```
Comment (bodzhang) [inline enclave/core/sgx/exception.c:411]: This TODO comment is obsolete, with the design of "only allow the simulated #PF in debug mode".

**Review by bodzhang — CHANGES_REQUESTED** [review]

File: enclave/core/sgx/exception.c (hunk 8 above)
Comment (mingweishih) [reply enclave/core/sgx/exception.c]: removed the comment

File: enclave/core/sgx/exception.c (hunk 7 above)
Comment (mingweishih) [reply enclave/core/sgx/exception.c]: Update to read the information passed in by the host. However, it seems the kernel does not allow the host to obtain the faulting address inside a user-mode signal handler; i.e., the address obtained from the signal handler on the host is always zero.

File: host/sgx/exception.c [hunk 9]
```diff
@@ -46,8 +47,10 @@ uint64_t oe_host_handle_exception(oe_host_exception_context_t* context)
 
         oe_exception_record_t oe_exception_record = {0};
 
-        /* TODO PRP: We need to save the exception information in
-         * oe_exception_record */
+        // ATTN: the faulting address is expected to be read from the
```
Comment (bodzhang) [inline host/sgx/exception.c:50]: In this case, the simulated #FP will not be very useful. I can't think of any realistic exception handler of #PF that does not need to check the faulting address.

Comment (mingweishih) [reply host/sgx/exception.c:50]: Should we remove the simulation code instead?

File: enclave/core/sgx/exception.c (hunk 7 above)
Comment (jxyang) [reply enclave/core/sgx/exception.c]: The kernel does pass a `mcontext` to the signal handler. In there the general purpose register  [` REG_CR2`](https://code.woboq.org/userspace/glibc/sysdeps/unix/sysv/linux/x86/sys/ucontext.h.html#REG_CR2) contains the faulty address of the sigsegv. We need to add plumbing to pass in this address from OE's signal handler on the host side to the signal handler in the enclave.

File: host/sgx/exception.c (hunk 9 above)
Comment (jxyang) [reply host/sgx/exception.c:50]: According to Peter, reporting faulty address at 0 is good enough for Java runtime. I think Java runtime depends on a SIGSEGV event, but not necessarily the address, to reach a safe point. In this regard, the simulated #FP event is still useful.

File: enclave/core/sgx/exception.c (hunk 7 above)
Comment (jxyang) [reply enclave/core/sgx/exception.c]: We have the dependency chain of:

 Release mode enclave -> openenclave.v4 -> this PR.

I think we should get this in quickly to move on to the other two issues.

Comment (mingweishih) [reply enclave/core/sgx/exception.c]: Just tested, the results is the same as `si_addr`.
I print the value of `REG_CR2` in the handler
```
    uint64_t faulting_address = (uint64_t)context->uc_mcontext.gregs[REG_CR2];
    printf("faulting address: 0x%lx\n", faulting_address);
```
and add an ocall to access address 0x8 at the host alone with the current two tests (access 0x8 inside the enclave and run `lgdt` instruction).

Here is the results
```
faulting address: 0x0
Test #PF passed
faulting address: 0x0
Test #GP passed
faulting address: 0x8
```

Comment (jxyang) [reply enclave/core/sgx/exception.c]: We need to verify when the faulty address is protected, i.e., inside the enclave, the kernel still report the correct address in si_addr or CR2.

Comment (mingweishih) [reply enclave/core/sgx/exception.c]: Does you mean accessing the enclave memory by the host? it does not trigger a #PF.
- Read -> get dummy value (-1)
- Write -> no effects

Comment (jxyang) [reply enclave/core/sgx/exception.c]: I mean changing the page permission on an enclave address to, for example, read only, and then writing to that address in the enclave code. Fundamentally this is the absolute dominate use case for #PF events in Mystikos.

Comment (mingweishih) [reply enclave/core/sgx/exception.c]: I can confirm that CR2/si_addr is not cleared if #PF is on an enclave memory.
I test with writing to an enclave code page.
```
enclave code addr: 0x7fc603003500
faulting address: 0x7fc603003000
```

Comment (mingweishih) [reply enclave/core/sgx/exception.c]: Let me update the comments and test cases accordingly.

File: host/sgx/exception.c (hunk 9 above)
Comment (bodzhang) [reply host/sgx/exception.c]: Even for Java/.Net runtime code that handles #PF on NULL reference, reporting faulting address of 0 when the faulting address is non-zero might be problematic. For other code, such as stack overflow detection, reporting incorrect faulting address will render the code logic useless, or even cause unexpected result.

**Review by bodzhang — APPROVED** [review]

**Pull request merged**